### ISSUE TITLE
Expanded developer API

### ISSUE THREAD

**Issue by alfonsoLeandro** [AlonsoAliaga/BetterBackpacks#4]
NOTICE THAT THIS IS JUST A REQUEST. FEATURE CAN NEVER BE ADDED.

Hey, I am a Spigot plugin developer, I devoleped an AutoPickup plugin that I published on Spigot, recently, a user of the plugin saw that it had support for another backpacks plugin and asked me if I could add support for your backpacks plugin, so I went on to check your developer API, however, I wasn't able to find any of the methods I needed, those are the following: check if player has 
any backpacks, get player's backpacks with their respective inventories.
Is there any possibilities you could add the methods I am missing in order to add support for your backpacks plugin?

**Is your feature request related to a problem? Please describe.**
No.

**Describe the solution you'd like**
Add "check if player has any backpack" and "get backpacks of player" methods

**Describe alternatives you've considered**
Maybe just a "get backpacks from player" and check if the returned list is empty?

**Additional context**
-

**Comment by AlonsoAliaga** [comment]
I doubt it's possible, considering that this is item based (not like one per player which can be kept in cache until disconnection).
These backpacks are loaded when right clicking the backpack and saved/serialized when closed the backpack. Hence why it's not possible to do that, that would mean looping items searcjing for backpacks and deserializing/serializing every time you want to add an item which i don't think is too performance friendly (as we use NBTAPI for nbt tags). Hence why there isn't a method for that, if they were virtual backpacks that are loaded on join and saved on left that would be easier
Hope you understood what I mean

**Comment by alfonsoLeandro** [comment]
Hmm, I see what you mean, so backpacks are stored in memory only when they are loaded, that is I suppose, every time a player opens the backpack, which they don't do nearly as often as picking up items. Could it be possible to instead add a "add item to play backpack" method that just serialized the item and added it to the items list?

**Comment by AlonsoAliaga** [comment]
Thats not possible either I think, the way items are stored are not just an serialized array of items, its more like an array of serialized items. Adding one item will always require deserializing/loading the whole backpack from the item, adding it to the inventory (while checking if there is space or not) and then serializing/saving it again (all that in milliseconds).
I also considered adding special backpacks to pick up specific materials, sadly as I mentioned above, if would be easy if it were loaded on cache as (virtual backpacks do) but this is an item-based backpack which makes it harder

**Comment by alfonsoLeandro** [comment]
I see, totally understand. I have 3 more questions:
1- Why does it work like this and not saving the backpacks in memory? which I'd think makes more sense, so I bet you have a reason for that, or did you just go with this solution?
2- Would you ever consider switching to saving player's backpacks when they have them in their inv and saving them to a file when they remove it from their inv?
3- How about adding the methods anyway and put a warning about performance issues on each of them? Maybe also making developers warn users of this, if they decide to include these methods in their projects?

**Comment by AlonsoAliaga** [comment]
1. Because thats not possible. As said above, if it were 1 per player (attaching it to player uuid) would be easy, but as players can have a lot backpacks on inventory, loading them is not an option (plus the fact they can drop, drag the backpack item)
2. No, will never, sorry. This plugin already works in a way, making such change will simply causr me issues when supporting pre and post versions (no, i will never tell them that old version is not supported and they MUST update to receive support). Additionally, recoding such big part of the system is not on my plans, thats basically a plugin recode, even with it, "saving when they remove it from inventory" is not that easy. There are many ways in which items are removed from inventory and i just cant track/liaten all of them (additionally, plugins that modify player's inventory will never be supported with that system as plugins simply setItem to null when moving items)
3. I was considering doing that, but as that requires a lot of work and hours of coding (i'm not considering doing it any time soon). I'm a college student, and I dont even have time to play. Adding new features (specially with such big changes) is not my priority. I honestly rather spending time playing or watching tv before going to bed (as i sleep early to wake up for classes) and I spend my day in front of the PC for classes (spending even more time in front of the PC coding something that will probably not be used as often, doesnt deserve the amount of time it requires to be coded) plus the fact that it can cause (performance) issues which (even with warnings) will affect my resource as people will be reporting them to me. If I ever have the time required for it, i'll consider it, however as I usually say, that doesnt mean I will do it. Some times i simply rather not doing it bc of time, losing interest, changes causes me more issues or simply not wanting to complicate even more my life (bc feature is hard to add).
Hope you get what I mean

**Comment by alfonsoLeandro** [comment]
I totally get what you mean, as I am a college student myself too, I totally understand your position, the answers you gave me are similar to the answers I thought you would give me, I just figured I'd ask. Thank you for considering it and thank you for your time. Keep up the good work

**Comment by joex92** [comment]
Hi, I'm the mentioned user... 
I haven't gone through the API, but would it be possible this situation then with the current API?
- Add items to the player's opened backpack
> As you stated, the backpack is loaded when right clicking it, so if it's already loaded, it should be possible to add items to it, right? The case scenario here would be: "I'm going to pick up a bunch of items on the floor, so I'm going to open the backpack and the overflowing items would go to the opened backpack inventory or the items would go directly to the back first and then to the inventory"

**Comment by AlonsoAliaga** [comment]
As mentioned above, that requires deserializing and serializing the whole backpack, which is not a simple process (internally). Already explained to alfonsoLeandro why it's not possible with the API. Backpacks are not "opened" if you haven't right click the item, backpacks are only "opened" when you are viewing the backpack inventory (thats the only moment where they are loaded), when player is not in that inventory backpack is not loaded. That wont change as that requires too much work, will cause me issues with the plugin and old versions and will probably introduce many issues i wont be able to fix, hence why this is an item based backpack item and has already his system, changing that (aka recoding) is not planned

**Comment by joex92** [comment]
I know... I understood that... my question is if it's possible when the backpack is being viewed by a player, so after the player right clicked it... obviously this would mean the player would not be doing anything else... but it is still a useful action...

**Comment by alfonsoLeandro** [comment]
That makes sense to me, what methods does the BackpackOpenEvent has?

**Comment by AlonsoAliaga** [comment]
Technically, yes it is, but you can't move with opened inventory.
Simply needs to check if opened inventory (holder) is instance of BetterBackpacksHolder, cast and check if getType() is 3 (not sure about the number, its used internally) and simply add the item as any other inventory.setItem (or even addItem, if available)

**Comment by AlonsoAliaga** [comment]
> That makes sense to me, what methods does the BackpackOpenEvent has?

Check it in your IDE, i dont remember, havent touch the api in a while. its 12:30 here, i'm sleeping now. If you want to ask something about development you can join my discord server so i can easily reply there

**Comment by joex92** [comment]
yes, I know that the player wouldn't be able to move... but that is better than nothing... the point is being able to "auto" pickup items to the backpack, in this case they would be "automatically" added to the backpack if the player inventory is full... that's my suggestion/workaround...

**Comment by alfonsoLeandro** [comment]
Great, thanks for letting me know how to do it, I will begin to work on it prolly tomorrow. Joex didn't mean to move with the inv open, he just wanted to pickup items dropped on the floor when opening a backpack, as if it increased the player's inventory size suddenly, I just need to get every entity in a maybe 1-1.5 block radius, despawn the entity if it is an item and add the ItemStack to the backpack. Not the best or cleanest solution, but it's a workaround, well thought @joex92

**Comment by AlonsoAliaga** [comment]
> yes, I know that the player wouldn't be able to move... but that is better than nothing... the point is being able to "auto" pickup items to the backpack, in this case they would be "automatically" added to the backpack if the player inventory is full... that's my suggestion/workaround...

If you plan to make a plugin to handle that, its ok, but that wont be added to this plugin, read my messages above stating why I'm not adding such features in this plugin

**Comment by alfonsoLeandro** [comment]
> Check it in your IDE, i dont remember, havent touch the api in a while. its 12:30 here, i'm sleeping now. If you want to ask something about development you can join my discord server so i can easily reply there

Will try tomorrow, it's 2am here haha, I'll make sure to join your discord tomorrow.




> If you plan to make a plugin to handle that, its ok, but that wont be added to this plugin, read my messages above stating why I'm not adding such features in this plugin

Oh no, I will add this feature to my already existing auto pickup plugin, not asking for anything 😄

**Comment by AlonsoAliaga** [comment]
> Great, thanks for letting me know how to do it, I will begin to work on it prolly tomorrow. Joex didn't mean to move with the inv open, he just wanted to pickup items dropped on the floor when opening a backpack, as if it increased the player's inventory size suddenly, I just need to get every entity in a 
> > Check it in your IDE, i dont remember, havent touch the api in a while. its 12:30 here, i'm sleeping now. If you want to ask something about development you can join my discord server so i can easily reply there
> 
> Will try tomorrow, it's 2am here haha, I'll make sure to join your discord tomorrow.
> 
> > If you plan to make a plugin to handle that, its ok, but that wont be added to this plugin, read my messages above stating why I'm not adding such features in this plugin
> 
> Oh no, I will add this feature to my already existing auto pickup plugin, not asking for anything 😄

Just make sure player dont use stack items plugin (as the amount of the itemstack is 1 but when picked up it adds the correct amount x100 for example)

**Comment by alfonsoLeandro** [comment]
> Just make sure player dont use stack items plugin (as the amount of the itemstack is 1 but when picked up it adds the correct amount x100 for example)

Will take it into consideration, thank you for letting me know.
